Amazon kms_facts module

willthames opened the pull request:
##### SUMMARY
Add a module for KMS facts

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
kms_facts

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel 0b784c65b1) last updated 2017/07/13 15:27:10 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION

Commit:
Amazon kms_facts module

Facts module for Amazon's Key Management Service

s-hertel commented on lib/ansible/modules/cloud/amazon/kms_facts.py:
```diff
@@ -0,0 +1,313 @@
+    description:
+      - A dict of filters to apply. Each dict item consists of a filter key and a filter value.
+        The filters aren't natively supported by boto3, but are supported to provide similar
+        functionality to other modules. Standard tag filters (C(tag-key), C(tag-value) and
```
Is C() only for files?
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

s-hertel commented on lib/ansible/modules/cloud/amazon/kms_facts.py:
```diff
@@ -0,0 +1,313 @@
+    description:
+      - A dict of filters to apply. Each dict item consists of a filter key and a filter value.
+        The filters aren't natively supported by boto3, but are supported to provide similar
+        functionality to other modules. Standard tag filters (C(tag-key), C(tag-value) and
```
Is C() only for files?
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

willthames replied:
`C()` is for option values. Here `filters` is the key so `tag-key` is part of the value (but also part of a key in the value). It's arguable, I'd say.

s-hertel replied:
I apparently cannot read.

s-hertel commented on lib/ansible/modules/cloud/amazon/kms_facts.py:
```diff
@@ -0,0 +1,313 @@
+        try:
+            tag_response = get_kms_tags_with_backoff(connection, key_id, **kwargs)
+            tags.extend(tag_response['Tags'])
+        except botocore.exceptions.ClientError as e:
```
It looks like the error response for keys that don't have the option of having tags and insufficient permissions for getting tags are the same. Maybe we shouldn't fail here if e.response['Error']['Code'] == 'AccessDeniedException' and just let the keys be empty. I'm not sure if there's a way of differentiating between the permissions and the acceptable key options for special keys.

s-hertel commented on lib/ansible/modules/cloud/amazon/kms_facts.py:
```diff
@@ -0,0 +1,313 @@
+            kwargs['Marker'] = tag_response['NextMarker']
+        else:
+            more = False
+    return boto3_tag_list_to_ansible_dict(tags)
```
I think this should be `return boto3_tag_list_to_ansible_dict(tags, 'TagKey', 'TagValue')` to get any tags back
http://boto3.readthedocs.io/en/latest/reference/services/kms.html#list_resource_tags

s-hertel commented on lib/ansible/modules/cloud/amazon/kms_facts.py:
```diff
@@ -0,0 +1,313 @@
+
+    all_keys = get_kms_facts(connection, module)
+    filtered = [key for key in all_keys if key_matches_filters(key, module.params['filters'])]
+    module.exit_json(keys=[camel_dict_to_snake_dict(key) for key in filtered])
```
The tag keys are being lowercased (maybe not a big deal, but I spent 10 minutes trying to figure out why filtering by them wasn't working).

willthames replied:
You've done more testing on the tag functionality than I have, sorry!

It's always best to do camel_to_snake before doing tag conversion, I should know this by now.

Commit:
kms_facts provide aliases

Return aliases for keys
Provide `alias` as a filter
Cope when tags can't be listed
Ensure everything is properly snake cased

Commit:
Rename kms_facts to aws_kms_facts

There may be conflicting KMS modules for other providers otherwise.

Commit:
Fix documentation, add aliases cache

Aliases are called many times, so add a cache

ryansb commented on lib/ansible/modules/cloud/amazon/aws_kms_facts.py:
```diff
@@ -0,0 +1,355 @@
+@AWSRetry.backoff(tries=5, delay=5, backoff=2.0)
+def get_kms_aliases_with_backoff(connection):
+    paginator = connection.get_paginator('list_aliases')
+    return paginator.paginate().build_full_result()['Aliases']
```
The other `get_kms_*` functions return the full response. Can you either return the full response here, or standardize on only returning the objects?

s-hertel commented:
shipit

Commit:
Reduce amount of info on deleted keys

Getting info on a key is costly (2s) per key, so reduce
info on deleted keys.

Commit:
Add policy information to facts

Commit:
aws_kms_facts version update

Fix ridiculously long RETURN line

Commit:
Remove dangerous-default-value from aws_kms_facts

willthames commented:
!needs_revision

s-hertel commented:
I think rebasing resets the ansibot's stale_review label process, and then it takes another [week](https://github.com/ansible/ansibullbot/blob/c665e5b0e00bff15bd68457883baccdfefd60a7e/ansibullbot/triagers/plugins/needs_revision.py#L359) to remove needs_revision and add stale_review.

willthames commented:
@jctanner mentioned that the bot won't remove needs_revision until @ryansb's rejection has been removed.

s-hertel commented:
bot_status
